circuit/src/commonMain/kotlin/com/slack/circuit/Overlay.kt

```diff
@@ -0,0 +1,185 @@
+ * is useful for one-off request/result flows such as:
+ * - Bottom sheets
+ * - Modals/dialogs
+ * - Tooltips
+ * - Full-screen takeover prompts
+ * - etc.
+ *
+ * The suspend [show] function is generically typed and can be suspended on to await the result of
+ * whatever overlay was launched.
+ *
+ * [currentOverlayData] is read-only and can be used to observe the current overlay data. This is
+ * generally intended to be used wherever the [OverlayHost] is provided via [LocalOverlayHost].
+ *
+ * In Android, this can be managed via the `ContentWithOverlays` composable function.
+ *
+ * To avoid accidentally capturing unnecessary state, it's recommended to create extension functions
+ * on `OverlayHost` that call [show] with the appropriate overlay and result type.
+ *
+ * ```
+ * private suspend fun OverlayHost.confirm(message: String): ConfirmationResult {
+ *   return show(ConfirmationOverlay(message))
+ * }
+ * ```
+ */
+@Stable
+public interface OverlayHost {
+  /**
+   * The current [OverlayHostData] or null if no overlay is currently showing.
+   *
+   * ```
+   * val overlayHostData by rememberUpdatedState(overlayHost.currentOverlayData)
+   * Box(modifier) {
+   *   content() // The regular content
+   *   key(overlayHostData) { overlayHostData?.let { data -> data.overlay.Content(data::finish) } }
+   * }
+   * ```
+   */
+  public val currentOverlayData: OverlayHostData<Any>?
+
+  /**
+   * Shows the given [overlay] and suspends until the overlay is finished with a [Result]. The
+   * overlay should _always_ signal a result (even if it's just something like `Result.Dismissed`)
+   * so that the [OverlayHost] can properly clear its [currentOverlayData].
+   *
+   * If no data is needed in a result, use [Unit] for the result type.
+   *
+   * This function should only be called from UI contexts and _not_ presenters, as overlays are a UI
+   * concern.
+   */
+  public suspend fun <Result : Any> show(overlay: Overlay<Result>): Result
+}
+
+/** Returns a remembered an [OverlayHost] that can be used to show overlays. */
+@Composable public fun rememberOverlayHost(): OverlayHost = remember { OverlayHostImpl() }
+
+private class OverlayHostImpl : OverlayHost {
+  /**
+   * Only one [Overlay] can be shown at a time. Since a suspending Mutex is a fair queue, this
+   * manages our message queue and we don't have to maintain one.
+   */
+  private val mutex = Mutex()
+
+  override var currentOverlayData: OverlayHostData<Any>? by mutableStateOf(null)
+    private set
+
+  override suspend fun <T : Any> show(overlay: Overlay<T>): T =
+    mutex.withLock {
+      try {
+        return suspendCancellableCoroutine { continuation ->
+          @Suppress("UNCHECKED_CAST")
+          currentOverlayData =
+            OverlayHostDataImpl(
+              overlay as Overlay<Any>,
+              continuation as CancellableContinuation<Any>
+            )
+        }
```

Comment:
I know we went over this but I spent a bunch of time trying to figure out how this was working. What was confusing me was that we were setting `currentOverlayData` in the coroutine block then nulling out the same prop in the finally block. I couldn't see how this was working and it was driving me crazy.

The piece of info I kept forgetting is that this method suspends and won't return _anything_ until something calls `Continuation.resume`